eng/native/configureplatform.cmake

```diff
@@ -327,19 +327,27 @@ else()
 endif()
 
 if(NOT CLR_CMAKE_HOST_ARCH_WASM)
-if(NOT CLR_CMAKE_TARGET_ANDROID) # Android requires PIC and CMake handles this so we don't need the check
-    # All code we build should be compiled as position independent
-    get_property(languages GLOBAL PROPERTY ENABLED_LANGUAGES)
-    if("CXX" IN_LIST languages)
-        set(CLR_PIE_LANGUAGE CXX)
-    else()
-        set(CLR_PIE_LANGUAGE C)
-    endif()
-    check_pie_supported(OUTPUT_VARIABLE PIE_SUPPORT_OUTPUT LANGUAGES ${CLR_PIE_LANGUAGE})
-    if(NOT MSVC AND NOT CMAKE_${CLR_PIE_LANGUAGE}_LINK_PIE_SUPPORTED)
-        message(WARNING "PIE is not supported at link time: ${PIE_SUPPORT_OUTPUT}.\n"
-                  "PIE link options will not be passed to linker.")
-    endif()
+    # Skip check_pie_supported call on Android as ld from llvm toolchain with NDK API level 21
+    # complains about missing linker flag `-no-pie` (while level 28's ld does support this flag,
+    # but since we know that PIE is supported, we can safely skip this redundant check).
+    if(NOT CLR_CMAKE_TARGET_ANDROID)
+        # All code we build should be compiled as position independent
+        get_property(languages GLOBAL PROPERTY ENABLED_LANGUAGES)
+        if("CXX" IN_LIST languages)
+            set(CLR_PIE_LANGUAGE CXX)
+        else()
+            set(CLR_PIE_LANGUAGE C)
+        endif()
+        check_pie_supported(OUTPUT_VARIABLE PIE_SUPPORT_OUTPUT LANGUAGES ${CLR_PIE_LANGUAGE})
+        if(NOT MSVC AND NOT CMAKE_${CLR_PIE_LANGUAGE}_LINK_PIE_SUPPORTED)
+            message(WARNING "PIE is not supported at link time: ${PIE_SUPPORT_OUTPUT}.\n"
+                      "PIE link options will not be passed to linker.")
+        endif()
+    endif(NOT CLR_CMAKE_TARGET_ANDROID)
+
     set(CMAKE_POSITION_INDEPENDENT_CODE ON)
-endif(NOT CLR_CMAKE_TARGET_ANDROID)
 endif(NOT CLR_CMAKE_HOST_ARCH_WASM)
+
+# Clear _FORTIFY_SOURCE=2, if set
+string(REPLACE "-D_FORTIFY_SOURCE=2 " "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+string(REPLACE "-D_FORTIFY_SOURCE=2 " "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
```

Comment:
This is a pretty ugly way of unsetting `_FORTIFY_SOURCE`. Can you try adding `#undef _FORTIFY_SOURCE` into the source files that are running into the issues instead?

Reply:
I prefer the way @am11 has done it. Otherwise we may need to keep adding that #undef to various places where the problem would pop. Moreover, it would be kind of lost in the sources.

Reply:
String replacing in CMake variables gives me shivers ;)

It seems like the `_FORITIFY_SOURCE` is only an issue during Debug builds because fortify requires optimization (-O) according to the error in the issue: https://github.com/dotnet/runtime/issues/2188

@am11 can you put the replacing under an `if (CMAKE_BUILD_TYPE STREQUAL Debug)` ?

Reply:
Actually, it seems that we could use `remove_definitions(-D_FORTIFY_SOURCE=2)` here and stay on the nicer side.

Reply:
String replacement was not my first choice either. When the C{XX}FLAGS are set in the environment we have very few options, `string replacement`, `parse to list remove and convert back to string` etc.
Release build is also affected by _FORTIFY_SOURCE=2.

Reply:
Tested with this delta:

```diff
diff --git a/eng/native/configureplatform.cmake b/eng/native/configureplatform.cmake
index 0d57c8fd00e..7ff0ae6db0e 100644
--- a/eng/native/configureplatform.cmake
+++ b/eng/native/configureplatform.cmake
@@ -349,5 +349,7 @@ if(NOT CLR_CMAKE_HOST_ARCH_WASM)
 endif(NOT CLR_CMAKE_HOST_ARCH_WASM)
 
 # Clear _FORTIFY_SOURCE=2, if set
-string(REPLACE "-D_FORTIFY_SOURCE=2 " "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
-string(REPLACE "-D_FORTIFY_SOURCE=2 " "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+message(WARNING "[BEFORE] CMAKE_C_FLAGS: ${CMAKE_C_FLAGS}\nCMAKE_CXX_FLAGS: ${CMAKE_CXX_FLAGS}"})
+remove_definitions(-D_FORTIFY_SOURCE=2)
+message(WARNING "[AFTER] CMAKE_C_FLAGS: ${CMAKE_C_FLAGS}\nCMAKE_CXX_FLAGS: ${CMAKE_CXX_FLAGS}"})
+message(FATAL_ERROR "intentional exit")
```

output:

```sh
CMake Warning at /runtime/eng/native/configureplatform.cmake:352 (message):
  [BEFORE] CMAKE_C_FLAGS: -g -O3 -feliminate-unused-debug-types -pipe -Wall
  -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector
  --param=ssp-buffer-size=32 -Wformat -Wformat-security -m64
  -fasynchronous-unwind-tables -Wp,-D_REENTRANT
  -ftree-loop-distribute-patterns -Wl,-z -Wl,now -Wl,-z -Wl,relro
  -fno-semantic-interposition -ffat-lto-objects -fno-trapping-math
  -Wl,-sort-common -Wl,--enable-new-dtags -mtune=skylake
  -Wa,-mbranches-within-32B-boundaries

  CMAKE_CXX_FLAGS: }
Call Stack (most recent call first):
  /runtime/eng/native/configuretools.cmake:1 (include)
  CMakeLists.txt:6 (include)


CMake Warning at /runtime/eng/native/configureplatform.cmake:354 (message):
  [AFTER] CMAKE_C_FLAGS: -g -O3 -feliminate-unused-debug-types -pipe -Wall
  -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector
  --param=ssp-buffer-size=32 -Wformat -Wformat-security -m64
  -fasynchronous-unwind-tables -Wp,-D_REENTRANT
  -ftree-loop-distribute-patterns -Wl,-z -Wl,now -Wl,-z -Wl,relro
  -fno-semantic-interposition -ffat-lto-objects -fno-trapping-math
  -Wl,-sort-common -Wl,--enable-new-dtags -mtune=skylake
  -Wa,-mbranches-within-32B-boundaries

  CMAKE_CXX_FLAGS: }
Call Stack (most recent call first):
  /runtime/eng/native/configuretools.cmake:1 (include)
  CMakeLists.txt:6 (include)


CMake Error at /runtime/eng/native/configureplatform.cmake:355 (message):
  intentional exit
```

if we plan to make sources fortify source compatible in future, then these two lines will go away. Otherwise, they can stay and do not cause any problem to the supported systems matrix (covered by 130 jobs run in CI).

Reply:
Hmm, I guess that's caused by the fact that the symbol is prefixed by `-Wp,` on the command line, which means "pass me to the preprocessor". So it is not part of the regular definitions.
That also means that your way of removal will leave `-Wp,` on the command line, doesn't it?

Reply:
@janvorli, it does, but `-Wp` does not affect the build breakage so I kept the change limited to what was needed to make the build green. This was the docker file which I posted and can be used to confirm the behavior being described: https://github.com/dotnet/runtime/issues/2188#issuecomment-600067624.

Reply:
Ok, that's fine.